File: modules/yacboard/pos/pos.go
package pos

type Pos int
type PosF byte

func FromXY(x, y int) Pos {
	if uint(x) >= Width || uint(y) >= Height {
		return -1
	}
	return Pos(x + y*Width)
}

func FromChars(chars string) Pos {
	if len(chars) < 2 {
		return -1
	}
	x := int(chars[0]) - 'a'
	y := Height + '0' - int(chars[1])
	return FromXY(x, y)
}

func (pos Pos) String() string {
	if uint(pos) >= FieldCount {
		return "-"
	}
	file := pos%Width + 'a'
	rank := Height - pos/Height + '0'
	return string([]byte{byte(file), byte(rank)})
}

func PToF(pos int) PosF {
	if pos < 0 {
		return 0
	}
	return PosF(pos%Width + 1 + (pos/Width+1)*WidthF)
}

func PToFb(pos byte) PosF {
	return PToF(int(pos))
}

func PToFp(pos Pos) PosF {
	return PToF(int(pos))
}

func FToP(pos PosF) int {
	if pos == 0 {
		return -1
	}
	return int(pos)%WidthF - 1 + (int(pos)/WidthF-1)*Width
}

File: dart/yacboard/lib/yacboard/pos.dart
import 'board_size.dart';

class Pos {
  static int fromXY(int x, int y) {
    if (x < 0 || x >= BoardSize.width || y < 0 || y >= BoardSize.height) return -1;
    return y * BoardSize.width + x;
  }

  static int fromChars(String c) {
    if (c.length < 2) return -1;
    int x = c.codeUnitAt(0) - 97; // 97 = 'a'
    int y = BoardSize.height + 48 - c.codeUnitAt(1); // 48 = '0'
    return fromXY(x, y);
  }

  static String asString(int pos) {
    if (pos < 0 || pos >= BoardSize.fieldCount) return "-";
    int file = (pos % BoardSize.width) + 97; // 97 = 'a'
    int rank = BoardSize.height - pos ~/ BoardSize.height + 48; // 48 = '0'
    return String.fromCharCodes([file, rank]);
  }
}
